source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,201 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
+    timer_ = dispatcher().createTimer(
+        [this, admin_handler] { this->onTimeout(admin_handler->attached_request_); });
+    timer_->enableTimer(std::chrono::milliseconds(timeout));
+  }
+}
+
+void AdminHandler::AttachedRequestBuffered::onTimeout(
+    std::weak_ptr<AttachedRequest> weak_attached_request) {
+  // Flush the buffer regardless of size
+  dispatcher().post([weak_attached_request] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      ENVOY_LOG(debug, "Timer Expiry after admin tap request completion");
+      return; // No attached request, flushed already
+    }
+    TraceBuffer* trace_buffer = attached_request->traceBuffer();
+
+    // if the trace buffer has already been flushed short circuit
+    if (trace_buffer->flushed()) {
+      return;
+    }
+
+    std::vector<envoy::data::tap::v3::TraceWrapper> buffer = trace_buffer->flush();
+
+    ENVOY_LOG(debug, "Timer Expiry, admin flushing buffered traces to response");
+
+    // Serialize writes to the stream
+    for (const auto& el : buffer) {
+      attached_request->streamMsg(el);
+    }
+    attached_request->endStream();
   });
 }
 
+void AdminHandler::AttachedRequest::endStream() {
+  Buffer::OwnedImpl output_buffer(std::string(""));
+  stream()->getDecoderFilterCallbacks().encodeData(output_buffer, true);
+}
+
+void AdminHandler::AttachedRequest::streamMsg(const Protobuf::Message& message, bool end_stream) {
+  std::string output_string;
+
+  switch (format()) {
+    PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
+  case envoy::config::tap::v3::OutputSink::JSON_BODY_AS_STRING:
+  case envoy::config::tap::v3::OutputSink::JSON_BODY_AS_BYTES:
+    output_string = MessageUtil::getJsonStringFromMessageOrError(message, true, true);
+    break;
+  case envoy::config::tap::v3::OutputSink::PROTO_BINARY_LENGTH_DELIMITED: {
+    Protobuf::io::StringOutputStream stream(&output_string);
+    Protobuf::io::CodedOutputStream coded_stream(&stream);
+    coded_stream.WriteVarint64(message.ByteSizeLong());
+    message.SerializeWithCachedSizes(&coded_stream);
+    break;
+  }
+  case envoy::config::tap::v3::OutputSink::PROTO_BINARY:
+  case envoy::config::tap::v3::OutputSink::PROTO_TEXT:
+    PANIC("not implemented");
+  }
+
+  Buffer::OwnedImpl output_buffer{output_string};
+  stream()->getDecoderFilterCallbacks().encodeData(output_buffer, end_stream);
+}
+
+AdminHandler::AttachedRequest* AdminHandler::AttachedRequest::createAttachedRequest(
+    AdminHandler* admin_handler, const envoy::admin::v3::TapRequest& tap_request,
+    Server::AdminStream* admin_stream) {
+  using ProtoOutputSink = envoy::config::tap::v3::OutputSink;
+
+  auto sink = tap_request.tap_config().output_config().sinks()[0];
+
+  switch (sink.output_sink_type_case()) {
+  case ProtoOutputSink::kBufferedAdmin:
+    return new AttachedRequestBuffered(admin_handler, tap_request, admin_stream);
```

Comment:
use std::make_shared rather than raw `new`. Looks like the call-site just assigns this anyway.